src/flp/types/fixedpoint_l2.rs

```diff
@@ -0,0 +1,816 @@
+        let gadget0_chunk_len = std::cmp::max(1, (len0 as f64).sqrt() as usize);
+        let mut gadget0_calls = len0 / gadget0_chunk_len;
+        if len0 % gadget0_chunk_len != 0 {
+            gadget0_calls += 1;
+        }
+
+        let len1 = entries;
+        let gadget1_chunk_len = std::cmp::max(1, (len1 as f64).sqrt() as usize);
+        let mut gadget1_calls = len1 / gadget1_chunk_len;
+        if len1 % gadget1_chunk_len != 1 {
+            gadget1_calls += 1;
+        }
+
+        Ok(Self {
+            bits_per_entry,
+            fi_bits_per_entry,
+            entries,
+            bits_for_norm,
+            range_01_checker: poly_range_check(0, 2),
+            norm_summand_poly,
+            phantom: PhantomData,
+
+            // range constants
+            range_norm_begin: entries * bits_per_entry,
+            range_norm_end: entries * bits_per_entry + bits_for_norm,
+
+            // configuration of parallel sum gadgets
+            gadget0_calls,
+            gadget0_chunk_len,
+            gadget1_calls,
+            gadget1_chunk_len,
+        })
+    }
+}
+
+impl<
+        T: Fixed,
+        F: FieldElement,
+        SPoly: ParallelSumGadget<F, PolyEval<F>>,
+        SBlindPoly: ParallelSumGadget<F, BlindPolyEval<F>>,
+    > Clone for FixedPointBoundedL2VecSum<T, F, SPoly, SBlindPoly>
+{
+    fn clone(&self) -> Self {
+        Self {
+            bits_per_entry: self.bits_per_entry,
+            fi_bits_per_entry: self.fi_bits_per_entry,
+            entries: self.entries,
+            bits_for_norm: self.bits_for_norm,
+            range_01_checker: self.range_01_checker.clone(),
+            norm_summand_poly: self.norm_summand_poly.clone(),
+            phantom: PhantomData,
+
+            // range constants
+            range_norm_begin: self.range_norm_begin,
+            range_norm_end: self.range_norm_end,
+
+            // configuration of parallel sum gadgets
+            gadget0_calls: self.gadget0_calls,
+            gadget0_chunk_len: self.gadget0_chunk_len,
+            gadget1_calls: self.gadget1_calls,
+            gadget1_chunk_len: self.gadget1_chunk_len,
+        }
+    }
+}
+
+impl<T, F, SPoly, SBlindPoly> Type for FixedPointBoundedL2VecSum<T, F, SPoly, SBlindPoly>
+where
+    T: Fixed + CompatibleFloat<F>,
+    F: FieldElement,
+    SPoly: ParallelSumGadget<F, PolyEval<F>> + Eq + 'static,
+    SBlindPoly: ParallelSumGadget<F, BlindPolyEval<F>> + Eq + 'static,
+{
+    const ID: u32 = 0xFFFF0000;
+    type Measurement = Vec<T>;
+    type AggregateResult = Vec<f64>;
+    type Field = F;
+
+    fn encode_measurement(&self, fp_entries: &Vec<T>) -> Result<Vec<F>, FlpError> {
+        // Convert the fixed-point encoded input values to field integers. We do
+        // this once here because we need them for encoding but also for
+        // computing the norm.
+        let integer_entries: Vec<_> = fp_entries.iter().map(|x| x.to_field_integer()).collect();
```

Comment:
You should be able to skip `collect()`, which would save a malloc. You might need a type hint elsewhere.

```suggestion
        let integer_entries: Vec<_> = fp_entries.iter().map(|x| x.to_field_integer());
```